src/workerd/io/worker.c++

```diff
@@ -3565,4 +3568,25 @@ kj::Promise<WorkerInterface::CustomEvent::Result>
   return inner->customEvent(kj::mv(event));
 }
 
+int32_t Worker::getNextPromiseContextTag() const {
+  // This calculation gives us 2,147,483,647 IoContexts before the counter rolls
+  // over, meaning that there is an ever so slight possibility of conflicts if a
+  // worker is busy enough. That's a large enough range that it's probably unlikely
+  // to be an issue?
+  //
+  // We use a value in the v8 SMI range currently because that's the easiest and
+  // most performant thing That Just Works when extending v8's internal JSPromise
+  // to include this additional information. We could use a uint32_t or uint64_t
+  // range value to expand the range but we'd have to either wrap that in a v8
+  // Value (which requires additional allocations) or or teach v8's garbage collector
+  // about a JSPromise with an additional uint32_t or uint64_t field, which is a
+  // much more complicated change.
+  static constexpr int32_t kSmiMax = v8::internal::kSmiMaxValue;
+  static constexpr int32_t kSmiMin = v8::internal::kSmiMinValue;
+  int32_t ret = (__atomic_fetch_add(&impl->promiseContextTag, 1, __ATOMIC_RELAXED) %
+                 kSmiMax) + kSmiMin;
```

Comment:
To get the full range I think this should be something like: `(counter % (kSmiMax - kSmiMin)) + kSmiMin)`